File: sdk/ts/__tests__/xlang-humaninloop.test.ts
// Cross-language conformance: human-in-loop scenario -- TypeScript (offline).

interface XlangHilEvent {
  kind: string
  run_id: string
  prompt?: string
  options?: string[]
  decision?: string
  output?: Record<string, unknown>
}

const XLANG_HIL_RUN_ID = 'xlh-ts'

function makeXlangHilEvents(runId: string): XlangHilEvent[] {
  return [
    { kind: 'started',            run_id: runId },
    { kind: 'decision_requested', run_id: runId, prompt: 'Please approve the draft', options: ['approve', 'reject'] },
    { kind: 'decision_received',  run_id: runId, decision: '{"approved":true}' },
    { kind: 'completed',          run_id: runId, output: { result: 'hil-ok' } },
  ]
}

describe('xlang human-in-loop scenario -- TypeScript', () => {
  const events = makeXlangHilEvents(XLANG_HIL_RUN_ID)

  it('started event is first', () => {
    expect(events[0].kind).toBe('started')
  })

  it('decision_requested comes before decision_received', () => {
    const decKinds = events.filter(e => e.kind.startsWith('decision')).map(e => e.kind)
    expect(decKinds).toEqual(['decision_requested', 'decision_received'])
  })

  it('decision is approved', () => {
    const received = events.find(e => e.kind === 'decision_received')!
    const dec = JSON.parse(received.decision!)
    expect(dec.approved).toBe(true)
  })

  it('prompt is non-empty', () => {
    const requested = events.find(e => e.kind === 'decision_requested')!
    expect(requested.prompt).toBeTruthy()
    expect(requested.options?.length).toBeGreaterThan(0)
  })

  it('completed event is last', () => {
    expect(events[events.length - 1].kind).toBe('completed')
  })

  it('run_id consistent across events', () => {
    for (const ev of events) {
      expect(ev.run_id).toBe(XLANG_HIL_RUN_ID)
    }
  })
})

File: sdk/go/ancora/xlang_humaninloop_test.go
package ancora_test

import (
	"encoding/json"
	"testing"
)

// Cross-language conformance: human-in-loop scenario -- Go.

type xlangHilEvent struct {
	Kind     string `json:"kind"`
	RunID    string `json:"run_id"`
	Prompt   string `json:"prompt,omitempty"`
	Decision string `json:"decision,omitempty"`
	Output   string `json:"output,omitempty"`
}

func makeXlangHilEvents(runID string) []xlangHilEvent {
	return []xlangHilEvent{
		{Kind: "started",            RunID: runID},
		{Kind: "decision_requested", RunID: runID, Prompt: "Please approve the draft"},
		{Kind: "decision_received",  RunID: runID, Decision: `{"approved":true}`},
		{Kind: "completed",          RunID: runID, Output: `{"result":"hil-ok"}`},
	}
}

func TestXlangGoHilStartedFirst(t *testing.T) {
	evs := makeXlangHilEvents("xlh-go")
	if evs[0].Kind != "started" {
		t.Fatalf("expected started, got %q", evs[0].Kind)
	}
}

func TestXlangGoHilRequestedBeforeReceived(t *testing.T) {
	evs := makeXlangHilEvents("xlh-go")
	var idx []int
	for i, e := range evs {
		if e.Kind == "decision_requested" || e.Kind == "decision_received" {
			idx = append(idx, i)
		}
	}
	if len(idx) != 2 || idx[0] >= idx[1] {
		t.Fatalf("requested must come before received, got indices %v", idx)
	}
}

func TestXlangGoHilDecisionIsApproved(t *testing.T) {
	evs := makeXlangHilEvents("xlh-go")
	var received xlangHilEvent
	for _, e := range evs {
		if e.Kind == "decision_received" {
			received = e
		}
	}
	var dec map[string]interface{}
	if err := json.Unmarshal([]byte(received.Decision), &dec); err != nil {
		t.Fatalf("unmarshal decision: %v", err)
	}
	if dec["approved"] != true {
		t.Fatalf("expected approved=true, got %v", dec["approved"])
	}
}

func TestXlangGoHilPromptNonEmpty(t *testing.T) {
	evs := makeXlangHilEvents("xlh-go")
	for _, e := range evs {
		if e.Kind == "decision_requested" && e.Prompt == "" {
			t.Fatal("prompt must be non-empty")
		}
	}
}

func TestXlangGoHilCompletedLast(t *testing.T) {
	evs := makeXlangHilEvents("xlh-go")
	if evs[len(evs)-1].Kind != "completed" {
		t.Fatalf("expected completed, got %q", evs[len(evs)-1].Kind)
	}
}
